#171380744 User sign up endpoint

the22mastermind opened the pull request:
#### What does this PR do?
Implement user sign up functionality
#### Description of Task to be completed?
* Implement endpoint unit tests
* Create sign up controller
* Create user model and migrations with sequelize
* Create helper functions, status codes, and messages.
* Install required libraries
* Create sign up route
* Initialize sign up route in `index.js`
* Implement jsDoc and endpoint documentation
#### How should this be manually tested?
1. Open your terminal and run `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend/`
2. run `cd stackup2-barefoot-backend` to go into the root directory of the project
3. run `git checkout -b ft-user-signup-jwt-171380744 origin/ft-user-signup-jwt-171380744` to pull this branch and switch to it
4. run `npm install` to install the dependencies
5. In `.env` file, configure these variables: 
```
NODE_ENV=test
TEST_DB_USERNAME=YOUR_USERNAME
TEST_DB_PASSWORD=YOUR_PASSWORD
TEST_DB_NAME=TEST_DB_NAME
TEST_DB_HOSTNAME=YOUR_HOSTNAME
TEST_DB_PORT=DB_PORT_NUMBER
JWT_KEY=SOME_VALUE
SIGN_UP_TOKEN_EXPIRES_IN=1h
```
6. run `npx sequelize-cli db:create` to create test db
7. run `npx sequelize-cli db:migrate` to save the user model in db
8. run `npm run test`
#### Any background context you want to provide?
* You need to install [Node.js](https://nodejs.org/en/) and [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) before testing.
* You need to have PostgeSQL database installed and configured on your computer
#### What are the relevant pivotal tracker stories?
[#171380741](https://www.pivotaltracker.com/story/show/171380744)
#### Screenshots (if appropriate)
N/A
#### Questions:

muchristian commented on src/routes/api/index.js:
```diff
@@ -0,0 +1,8 @@
+
+const router = express.Router();
+
+router.use('/v1/auth', authentication);
```
may be better if you try to keep the version(v1) in environment variable and codes may look like this `const version= process.env.Version || 'v1'`. you'll be using version variable frequently as you want without hard code it every time as you did it. means that you can replace `v1` with version variable.

descholar-ceo replied:
I don't understand why you keep using this `v1`, @muchristian @the22mastermind

muchristian commented on src/controllers/auth.js:
```diff
@@ -0,0 +1,51 @@
+        password,
+        gender,
+        address,
+      } = req.body;
```
you forgot to write some attribute should be there, for instance `passportId, addressId instead of address, isVerified, rememberMe, createdAt and updatedAt`.

the22mastermind replied:
I think the other attributes would be more appropriate for edit/update profile which is another story. For sign up we use the minimum required to create an account, then next step we edit/update profile.

the22mastermind committed:
feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

Resolve conflicts

clean up index.js

Remove api version in route

--amend

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

Resolve conflicts

clean up index.js

Remove api version in route

--amend

feature(signup):Implement signup endpoint [Finishes #171380744]

feature(signup):Implement signup endpoint [Finishes #171380744]

Resolve conflicts

clean up index.js

Remove api version in route

--amend

Clean up after resolving conflicts

Code refactoring

Camel case fix

Camel case fix

Remove http in swagger.json

since9629 commented:
@the22mastermind could you resolve conflicts in this PR

ricahcyuzuzo reviewed (commented):
Solve the conflicts in this PR and solve hound issues!

ricahcyuzuzo reviewed (commented):
Hey @the22mastermind, Why does the Travis is not building well? or it is not configured the Postgres?
If it is the problem, we have to wait for the fix!

the22mastermind commented:
Yes, the database is not yet configured that's why the tests are failing. The person who setup travis should fix this asap.

Muhire-Josue commented on tests/auth.test.js:
```diff
@@ -0,0 +1,53 @@
+      .request(server)
+      .post('/api/auth/signup')
+      .send({
+        firstname: 'John',
```
Can you create a `data` folder inside the test folder, create a file called `mockData` that returns an array of objects, then you can move this object inside that array and call it here. like `mockData[0]`

Muhire-Josue commented on tests/auth.test.js:
```diff
@@ -0,0 +1,53 @@
+      .request(server)
+      .post('/api/auth/signup')
+      .send({
+        firstname: 'John',
```
Same thing here, do `mockData[1]`

Muhire-Josue reviewed (changes requested):
could you rename this file `src/controllers/auth.js` to something more descriptive, like `src/controllers/userAuthentication.js`

Muhire-Josue reviewed (approved):
LGTM, great work @the22mastermind

ricahcyuzuzo reviewed (approved):
Hey, @the22mastermind You should also try to solve the coverage problems and code climate.

Ugizwenayo-Divine reviewed (commented):
You did a great job in implementing everything, but there is a test which is not passing as expected 
check this and correct that, please
![Capture](https://user-images.githubusercontent.com/52660542/76146907-585c4380-603b-11ea-897a-bf0aafbbbe8d.PNG)

descholar-ceo commented on src/database/models/user.js:
```diff
@@ -1,10 +1,17 @@
+    isVerified: DataTypes.BOOLEAN
   }, {});
-  User.associate = (models) => {
+  user.associate = (models) => {
```
could you remove this line too? We saw that we don't need it so far. I suggest this because it decreasing functions coverage

descholar-ceo reviewed (commented):
Well done, remember to make all variable camelCase

descholar-ceo reviewed (approved):
LGTM
